pubtools/_pulp/tasks/delete.py

```diff
@@ -0,0 +1,709 @@
+
+    repo = attr.ib()
+    """The repo which where content was cleared."""
+
+    content = attr.ib()
+    """The content that was cleared from the repo"""
+
+
+@attr.s
+class RemoveUnitItem(object):
+    """Represents a unit with destination repos"""
+
+    unit = attr.ib()
+    """A Pulp unit"""
+
+    repos = attr.ib()
+    """The list of repos from where the unit to be removed"""
+
+
+class Delete(PulpClientService, CollectorService, Publisher, PulpTask):
+    def __init__(self, *args, **kwargs):
+        super(Delete, self).__init__(*args, **kwargs)
+        self.to_await = []
+
+    def add_args(self):
+        super(Delete, self).add_args()
+
+        self.add_publisher_args(self.parser)
+
+        self.parser.add_argument(
+            "--repo",
+            help="remove content from these comma-seperated repositories ",
+            type=str,
+            action=SplitAndExtend,
+            split_on=",",
+        )
+
+        self.parser.add_argument(
+            "--advisory",
+            help="remove packages in the advisory",
+            type=str,
+        )
+
+        self.parser.add_argument(
+            "--file",
+            help="remove these comma-separated rpm, srpm, modulemd or iso file(s)",
+            type=str,
+            action=SplitAndExtend,
+            split_on=",",
+        )
+
+        self.parser.add_argument(
+            "--signing-key",
+            help="remove the content with these signing-keys(s)",
+            type=str,
+            action=SplitAndExtend,
+            split_on=",",
+        )
+
+    def fail(self, *args, **kwargs):
+        LOG.error(*args, **kwargs)
+        sys.exit(30)
+
+    def run(self):
+        repos = {}
+        rpm_f = f_return()
+        file_f = f_return()
+        module_f = f_return()
+        adv_f = f_return()
+        signing_keys = self.args.signing_key
+        repo_names = self.args.repo
+        advisory = self.args.advisory
+
+        if not (self.args.file or self.args.advisory):
+            self.fail("One of --file or --advisory is required")
+
+        # separates files into rpms, isos and modules
+        rpms, files, modules, _ = self.separate_files(self.args.file or [])
+
+        # collect skipped items
+        # TODO
+
+        # delete rpms
+        if rpms:
+            rpm_f = self.delete_rpms(rpms, signing_keys, repo_names)
+
+        # delete files
+        if files:
+            file_f = self.delete_files(files, repo_names)
+
+        # delete modules
+        if modules:
+            module_f = self.delete_modules(modules, signing_keys, repo_names)
+
+        # delete packages from advisory
+        if advisory:
+            adv_f = self.delete_from_advisory(advisory, repo_names)
+
+        # wait for the futures to resolve
+        # each future returns a list of futures for content removal on each repo
+        fs = f_zip(rpm_f, file_f, module_f, adv_f)
+        fs = [f for sublist in fs.result() if sublist for f in sublist]
```

Comment:
I find this tricky to read.  Would it be simpler if you initialized the futures like rpm_f as `f_return([])` instead of just `f_return()` ?  If methods are designed to return a list, you may as well use empty lists to represent "nothing was removed" rather than None.

Reply:
Ack